File: ios/Classes/TiImagepickerModule.m
/**
 * titanium-ios-imagepicker
 *
 * Created by Your Name
 * Copyright (c) 2018 Your Company. All rights reserved.
 */

#import "TiImagepickerModule.h"

#import "TiApp.h"
#import "TiBase.h"
#import "TiHost.h"
#import "TiUtils.h"

@implementation TiImagepickerModule

#pragma mark Internal

- (id)moduleGUID
{
  return @"680ebae9-5f3a-4a1d-9afc-31183b24157a";
}

- (NSString *)moduleId
{
  return @"ti.imagepicker";
}

#pragma mark Lifecycle

- (void)startup
{
  [super startup];
  DebugLog(@"[DEBUG] %@ loaded", self);
}

#pragma Public APIs

- (void)openGallery:(id)args
{
  ENSURE_SINGLE_ARG(args, NSDictionary);
  
  NSUInteger maxImageSelection = [TiUtils intValue:args[@"maxImageSelection"] def:1000];
  NSString *doneButtonTitle = [TiUtils stringValue:args[@"doneButtonTitle"]];
  NSString *cancelButtonTitle = [TiUtils stringValue:args[@"cancelButtonTitle"]];
  NSString *noCameraTitle = [TiUtils stringValue:args[@"noCameraTitle"]];
  NSString *noImagesTitle = [TiUtils stringValue:args[@"noImagesTitle"]];
  NSString *settingsTitle = [TiUtils stringValue:args[@"settingsTitle"]];
  NSUInteger columnCount = [TiUtils intValue:args[@"columnCount"] def:3];
  BOOL recordLocation = [TiUtils boolValue:args[@"recordLocation"] def:NO];

  _currentCallback = (KrollCallback *)args[@"callback"];

  Configuration *imagePickerConfig = [[Configuration alloc] init];
  imagePickerConfig.doneButtonTitle = doneButtonTitle;
  imagePickerConfig.cancelButtonTitle = cancelButtonTitle;
  imagePickerConfig.noCameraTitle = noCameraTitle;
  imagePickerConfig.noImagesTitle = noImagesTitle;
  imagePickerConfig.settingsTitle = settingsTitle;
  imagePickerConfig.recordLocation = recordLocation;

  ImagePickerController *imagePicker = [[ImagePickerController alloc] initWithConfiguration:imagePickerConfig];
  
  imagePicker.delegate = self;
  
  TiThreadPerformOnMainThread(^{
    [TiApp.app showModalController:imagePicker animated:YES];
  }, NO);
}

- (void)cancelButtonDidPress:(ImagePickerController * _Nonnull)imagePicker {
  [_currentCallback call:@[@{ @"success": @NO, @"cancel": @YES, @"images": @[] }] thisObject:self];
  [TiApp.app hideModalController:imagePicker animated:YES];
}

- (void)doneButtonDidPress:(ImagePickerController * _Nonnull)imagePicker images:(NSArray<UIImage *> * _Nonnull)images {
  NSMutableArray<TiBlob *> *mappedPhotos = [NSMutableArray arrayWithCapacity:images.count];
  
  [images enumerateObjectsUsingBlock:^(UIImage * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
    [mappedPhotos addObject:[[TiBlob alloc] _initWithPageContext:self.pageContext andImage:obj]];
  }];
  
  [_currentCallback call:@[@{ @"success": @YES, @"cancel": @NO, @"images": mappedPhotos }] thisObject:self];
  [TiApp.app hideModalController:imagePicker animated:YES];
}

- (void)wrapperDidPress:(ImagePickerController * _Nonnull)imagePicker images:(NSArray<UIImage *> * _Nonnull)images {
  [TiApp.app hideModalController:imagePicker animated:YES];
  DebugLog(@"[WARN] Wrapper pressed");
}

@end

File: ios/Classes/TiImagepickerModule.swift
//
//  TiImagepickerModule.swift
//  titanium-imagepicker
//
//  Created by Hans Knöchel
//  Copyright (c) 2019 Hans Knöchel. All rights reserved.
//

import UIKit
import TitaniumKit
import YPImagePicker

@objc(TiImagepickerModule)
class TiImagepickerModule: TiModule {

  public let testProperty: String = "Hello World"
  
  func moduleGUID() -> String {
    return "8477e94d-ea5f-4fe6-8f6a-1e6bacbda2d7"
  }
  
  override func moduleId() -> String! {
    return "ti.imagepicker"
  }

  override func startup() {
    super.startup()
    debugPrint("[DEBUG] \(self) loaded")
  }
  
  @objc(openGallery:)
  func openGallery(arguments: Array<Any>?) {
    guard let arguments = arguments, let options = arguments[0] as? [String: Any] else { return }
    guard let callback: KrollCallback = options["callback"] as? KrollCallback else { return }

    var config = YPImagePickerConfiguration()

    // Some hardcoded values that may become configurable in the future
    config.showsFilters = false
    config.startOnScreen = .library
    config.screens = [.library, .photo]
    config.shouldSaveNewPicturesToAlbum = false

    // General (optional) config
    config.library.numberOfItemsInRow = options["columnCount"] as? Int ?? 3

    if options["tintColor"] != nil {
      config.colors.tintColor = TiUtils.colorValue(options["tintColor"])!.color
    }

    if options["maxImageSelection"] != nil {
      config.library.maxNumberOfItems = options["maxImageSelection"] as? Int ?? 99
    }

    if options["doneButtonTitle"] != nil {
      config.wordings.done = options["doneButtonTitle"] as! String
    }

    if options["cancelButtonTitle"] != nil {
      config.wordings.cancel = options["cancelButtonTitle"] as! String
    }

    if options["nextButtonTitle"] != nil {
      config.wordings.next = options["nextButtonTitle"] as! String
    }

    if options["cameraTitle"] != nil {
      config.wordings.cameraTitle = options["cameraTitle"] as! String
    }

    if options["libraryTitle"] != nil {
      config.wordings.libraryTitle = options["libraryTitle"] as! String
    }

    if options["albumsTitle"] != nil {
      config.wordings.albumsTitle = options["albumsTitle"] as! String
    }

    if options["capturePhotoImage"] != nil {
      config.icons.capturePhotoImage = TiUtils.image(options["capturePhotoImage"], proxy: self)
    }

    let picker = YPImagePicker(configuration: config)

    picker.didFinishPicking { [unowned picker] items, cancelled in
      if cancelled {
        callback.call([["success": false, "cancel": true, "images": []]], thisObject: self)
        picker.dismiss(animated: true, completion: nil)
        return
      }
  
      var images: [TiBlob] = []

      for item in items {
        switch item {
        case .photo(let photo):
          images.append(self.blob(from: photo.image))
        case .video(_):
          print("[WARN] Videos are not handled so far")
        }
      }
  
      callback.call([["images": images, "success": true]], thisObject: self)
      picker.dismiss(animated: true, completion: nil)
    }
 
    guard let controller = TiApp.controller(), let topPresentedController = controller.topPresentedController() else {
      print("[WARN] No window opened. Ignoring gallery call …")
      return
    }

    topPresentedController.present(picker, animated: true, completion: nil)
  }
  
  private func blob(from image: UIImage) -> TiBlob {
    return TiBlob()._init(withPageContext: self.pageContext, andImage: image)
  }
}
